### PR TITLE
Feature- add apis based on validation result

### PR DESCRIPTION
**Pull request by rshaiju**
Following are addressed

-  Add an new api if it doesn't exist
-  Add a new version if one or more versions already exist

### CODE REVIEW

File: src/Kmd.Logic.Gateway.Automation/Client/ValidatePublishingRequest.cs [hunk 1]
```diff
@@ -17,5 +17,16 @@ public ValidatePublishingRequest(Guid providerId, IEnumerable<ApiValidationModel
...
 
         public IEnumerable<ProductValidationModel> Products { get; }
+
+        public void Dispose()
+        {
+            if (this.Apis != null)
+            {
+                foreach (var api in this.Apis)
+                {
+                    api.OpenApiSpec?.Dispose();
```
Comment (jaceks2106) [inline src/Kmd.Logic.Gateway.Automation/Client/ValidatePublishingRequest.cs:27]: I believe that they are disposed once they are sent with http call

File: src/Kmd.Logic.Gateway.Automation/Kmd.Logic.Gateway.Automation.csproj [hunk 2]
```diff
@@ -8,6 +8,10 @@
     <Description>Library for KMD Logic Gateway Automation</Description>
   </PropertyGroup>
 
+  <ItemGroup>
```
Comment (jaceks2106) [inline src/Kmd.Logic.Gateway.Automation/Kmd.Logic.Gateway.Automation.csproj:11]: Is it necessary?

File: src/Kmd.Logic.Gateway.Automation/Publish.cs [hunk 3]
```diff
@@ -162,6 +170,71 @@ private async Task UpdateProduct(IGatewayClient client, Guid subscriptionId, Gui
...
+        {
+            Guid? apiVersionSetId = existingApis.FirstOrDefault(existingApi => existingApi.Name == api.Name)?.ApiVersionSetId;
+
+            var productIds = apiVersion.ProductNames.Select(productName => allProducts.SingleOrDefault(product => string.Compare(product.Name, productName, comparisonType: StringComparison.OrdinalIgnoreCase) == 0).Id).ToList();
+            using var logo = new FileStream(path: Path.Combine(folderPath, apiVersion.ApiLogoFile), FileMode.Open);
+            using var document = new FileStream(path: Path.Combine(folderPath, apiVersion.ApiDocumentation), FileMode.Open);
+            using var openApiSpec = new FileStream(path: Path.Combine(folderPath, apiVersion.OpenApiSpecFile), FileMode.Open);
+
+            var response = await client.CustomCreateApiAsync(
+            subscriptionId: subscriptionId,
```
Comment (jaceks2106) [inline src/Kmd.Logic.Gateway.Automation/Publish.cs:216]: Wrong intendation

**Review by jaceks2106 — CHANGES_REQUESTED** [review]

File: src/Kmd.Logic.Gateway.Automation/Client/ValidatePublishingRequest.cs (hunk 1 above)
Comment (rshaiju) [reply src/Kmd.Logic.Gateway.Automation/Client/ValidatePublishingRequest.cs:27]: No. We were getting "file already open" error in the api publishing code when it tried to open the file for reading. Then we realized the stream opened by the validator was not getting closed. Hence doing it explicitly

File: src/Kmd.Logic.Gateway.Automation/Kmd.Logic.Gateway.Automation.csproj (hunk 2 above)
Comment (rshaiju) [reply src/Kmd.Logic.Gateway.Automation/Kmd.Logic.Gateway.Automation.csproj]: Fixed

File: src/Kmd.Logic.Gateway.Automation/Publish.cs (hunk 3 above)
Comment (rshaiju) [reply src/Kmd.Logic.Gateway.Automation/Publish.cs]: Corrected

File: src/Kmd.Logic.Gateway.Automation/Client/ValidatePublishingRequest.cs (hunk 1 above)
Comment (jaceks2106) [reply src/Kmd.Logic.Gateway.Automation/Client/ValidatePublishingRequest.cs:27]: So how is it possible that on calling ValidatePublishing I didn't have to dispose them?

File: src/Kmd.Logic.Gateway.Automation/Kmd.Logic.Gateway.Automation.csproj [hunk 4]
```diff
@@ -8,7 +8,7 @@
     <Description>Library for KMD Logic Gateway Automation</Description>
   </PropertyGroup>
 
-  <ItemGroup>
+	<ItemGroup>
```
Comment (jaceks2106) [inline src/Kmd.Logic.Gateway.Automation/Kmd.Logic.Gateway.Automation.csproj:11]: Unnecessary intend

File: src/Kmd.Logic.Gateway.Automation/Client/ValidatePublishingRequest.cs (hunk 1 above)
Comment (jaceks2106) [reply src/Kmd.Logic.Gateway.Automation/Client/ValidatePublishingRequest.cs:27]: Ok, I got it right now. But test then if validation call still works ;)

File: src/Kmd.Logic.Gateway.Automation.Tool/Kmd.Logic.Gateway.Automation.Tool.csproj [hunk 5]
```diff
@@ -1,24 +1,24 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
-	<PropertyGroup>
```
Comment (jaceks2106) [inline src/Kmd.Logic.Gateway.Automation.Tool/Kmd.Logic.Gateway.Automation.Tool.csproj:3]: Why this file is changed?

File: src/Kmd.Logic.Gateway.Automation/Kmd.Logic.Gateway.Automation.csproj (hunk 4 above)
Comment (jaceks2106) [inline src/Kmd.Logic.Gateway.Automation/Kmd.Logic.Gateway.Automation.csproj:11]: Why this file is changed?

File: sample/Kmd.Logic.Gateway.Automation.Sample/Publish/publish.yml [hunk 6]
```diff
@@ -61,15 +61,15 @@ Apis:
...
       ApiDocumentation: ./apis/nexus3/markdown-sample.md
       OpenApiSpecFile: ./apis/nexus3/nexus-v1.json
       Published: true
-      ProductNames: [nexus-2]
+      ProductNames: [nexusxqk-2]
       PolicyXmlFile: ./apis/nexus3/policies.xml
       BackendLocation: http://nexus3
-   Path: nexus3
\ No newline at end of file
+   Path: nexusxqk3
```
Comment (jaceks2106) [inline sample/Kmd.Logic.Gateway.Automation.Sample/Publish/publish.yml:75]: Maybe let's not change those paths/names each time we want to test. We should clear our provider or do not push those changes.

**Review by jaceks2106 — DISMISSED** [review]

**Review by jaceks2106 — APPROVED** [review]

File: src/Kmd.Logic.Gateway.Automation/Kmd.Logic.Gateway.Automation.csproj [hunk 7]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-	  <TargetFramework>netstandard2.0</TargetFramework>
+          <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (jaceks2106) [inline src/Kmd.Logic.Gateway.Automation/Kmd.Logic.Gateway.Automation.csproj:4]: Oops?

**Review by jaceks2106 — CHANGES_REQUESTED** [review]

**Review by NakulSeerpu — APPROVED** [review]
LGTM

**Review by jaceks2106 — APPROVED** [review]

**Pull request merged**